### PR TITLE
Add script for importing variants to Moka

### PR DESCRIPTION
**Pull request by andyb3**
This change is [Reviewable](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3)

### CODE REVIEW

**Review by NMNS93 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3)* status: 0 of 5 files reviewed, 12 unresolved discussions (waiting on @aledj2, @andyb3, @NMNS93, and @woook)

___
*[100k_tier_1_2_vars.py, line 217 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQXZ7c-GczAUQpdoXX:-LjQXZ7c-GczAUQpdoXY:bq77hbw) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_tier_1_2_vars.py#L217)):*
> ```Python
>     ir_json = get_ir_json(args.ir_id.split('-')[0], args.ir_id.split('-')[1])
>     # Capture the genome assembly
>     assembly = ir_json.get('assembly')
> ```

Does there need to be a comment to say why we are using dict.get() instead of dict['key'] here?

It made me pause but we don't have any guidelines on this so I'm just double-checking.
___
*[100k_tier_1_2_vars.py, line 229 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQXuL1Dl40dZ8OHEvf:-LjQXuL1Dl40dZ8OHEvg:bj61kzo) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_tier_1_2_vars.py#L229)):*
> ```Python
>                 gt = zygosity_to_vcf(variant_call.zygosity)
>         # The other details needed are stored in 'variantCoordinates' 
>         # NOTE the assembly recorded in variantCoordinates is often wrong (!!!) so don't use, see GEL-42665
> ```

I'm guessing that we capture the genome assembly from the json rather than variantCoordinates. If so, this coomment should note that we use that instead?

Also is GEL-42665 a document or a ticket? I would suggest keeping this info in a readme/somewhere else unless it's a persistent document that can be referred to. My bad if it is!
___
*[100k_variant_import.py, line 8 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQYXseBTsfGc1ctE_f:-LjQYXseBTsfGc1ctE_g:b-60j8e6) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_import.py#L8)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> Requirements:
>     Python 2.7
>     paramiko
>     pyodbc
> ```
</details>


Could we remove this? Import statement tells me more and the shebang at the top could indicate python2.7

Standard is a requirements.txt file created with pip which we could consider making in the future.
___
*[100k_variant_import.py, line 163 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQcgv6CBTwNr-a0U8X:-LjQcgv6CBTwNr-a0U8Y:b-3dxnl0) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_import.py#L163)):*
> ```Python
> 
> 
>     def insert_to_moka(self, variant):
> ```

There's lots happening in this function that isn't inserting the variant into Moka. It looks like the dictionary being passed to this function requires a few 'cleaning' operations before it can go in.

Can we break those out into a separate function or even class, so that the variant passed to this is ready for insert?
___
*[100k_variant_import.py, line 167 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQbu4f4uXfE1NUWM9R:-LjQbu4f4uXfE1NUWM9S:b6avb47) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_import.py#L167)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
>         Args:
>             variant = dictionary in the format output from get_additional_info() function
>         Returns:
>             status string, indicating outcome of function (e.g. imported, skipped)
> ```
</details>


Suggestion for consistency/clarity. I use the google python style guide where I can:
```
Args:
    variant(dict): Variant data output from get_additional_info().
Returns:
    status(str): Indicates outcome of variant import into moka. Examples: imported, skipped.
```
___
*[100k_variant_import.py, line 172 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQd4k03knLEPkCcR2O:-LjQd4k03knLEPkCcR2P:beipvh2) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_import.py#L172)):*
> ```Python
>         """
>         # Convert chromosomes to Moka chromosome IDs
>         if variant['chr37']:
> ```

What's the difference between chr37/chr38 and pos37/pos38? May be worth a comment as I'm not sure what's happening
___
*[100k_variant_import.py, line 344 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQde8O9VHI3tSpdpPq:-LjQde8O9VHI3tSpdpPr:bseblpw) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_import.py#L344)):*
> ```Python
>     # The transcript field contains details for each transcript returned from API
>     # Store transcript annotations in list of dictionaries
>     # Separate the transcripts by splitting on comma, then capture the individual transcript annotations by splitting on semi-colon
> ```

Small thing but please could we move this comment down to above the for loop
___
*[100k_variant_import.py, line 432 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQe3LF2W43ke9WHVXp:-LjQe3LF2W43ke9WHVXq:b-tx124a) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_import.py#L432)):*
> ```Python
>         try:
>             variants_annotated.append(get_additional_info(variant))
>         except:
> ```

 I have seen catching all exceptions like this as a red flag before. Can it be avoided? I.e. Are the specific exceptions to catch?
___
*[100k_variant_import.py, line 452 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQeUFNDJnzRgQF7zf9:-LjQeUFNDJnzRgQF7zfA:b3nl7or) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_import.py#L452)):*
> ```Python
>             else:
>                 failed_import.append(variant['submitted_variant'])
>         except:
> ```

As with previous comment. And this is also done in the else statement above?
___
*[100k_variant_validator.py, line 49 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQgnyf-JfwiKZLZL_1:-LjQgnyg1kBKIM1LfB5M:b-thyghs) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_validator.py#L49)):*
> ```Python
>     Represents a genetic variant.
>     """
>     def __init__(self):
> ```

Should each variant class be initialised with the arguments that are passed to get_info?

Is get info to be used repeatedly on multiple variants? As we have all these self variables I didn't think so? Maybe the docstring can explain a bit more
___
*[100k_variant_validator.py, line 61 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQfNpPBPUo3oluqhd5:-LjQfNpPBPUo3oluqhd6:b-riya7n) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_validator.py#L61)):*
> ```Python
>         self.var_val_version = ''
> 
>     def get_info(self, build, chr, pos, ref, alt):
> ```

This was quite hard to follow. Potentially it could be broken up into:
* A function that gets the json
* A function that processes jsons with intergenic flags
* A function that processes jsons with gene_variant flags

get_info() could use the json flag to call the function you want. There are other ways but a dict can work too:
```python
flag_parser = {'intergenic': self.parse_intergenic, 'gene_variant': self.parse_gene_variant}
parser = flag_parser[json_flag]
parser(r.json)
```
___
*[README.md, line 111 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQhPJO8CsP1tKZURvx:-LjQhPJO8CsP1tKZURvy:beyyau2) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/README.md#L111)):*
> ```Markdown
> This script requires access to Moka so must be run from a trust Windows desktop with access to `\\gstt.local\apps\Moka\Files\Software\`. It uses the [`ssh_n_run.py`](https://github.com/moka-guys/ssh_to_genapp) script to run code remotely on the trust linux server.
> 
> Requirements:
> ```

Requirements.txt file might be better as suggested but if not using pip I understand listing here at least!

**Review by NMNS93 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3)* status: 0 of 5 files reviewed, 13 unresolved discussions (waiting on @aledj2, @andyb3, @NMNS93, and @woook)

___
*[README.md, line 117 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQh_KxEAOE0CEYqJzM:-LjQh_KyA40yiyB6yzpC:bbex3h0) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/README.md#L117)):*
> ```Markdown
> * pyodbc
> 
> The connection details for Moka  `config.ini` in the same directory of the script. See `example_config.ini` for format.
> ```

'are in'

**Review by andyb3 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3)* status: 0 of 5 files reviewed, 13 unresolved discussions (waiting on @aledj2, @andyb3, @NMNS93, and @woook)

___
*[100k_tier_1_2_vars.py, line 217 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQXZ7c-GczAUQpdoXX:-LjR0DsF93jH7aE_KBkr:bx5m80q) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_tier_1_2_vars.py#L217)):*
<details><summary><i>Previously, NMNS93 (Nana Mensah) wrote…</i></summary><blockquote>

Does there need to be a comment to say why we are using dict.get() instead of dict['key'] here?

It made me pause but we don't have any guidelines on this so I'm just double-checking.
</blockquote></details>

I've just changed it to use standard dict lookup. Think I'd just copied this from some other code, but can't see any reason to use the .get method
___
*[100k_tier_1_2_vars.py, line 229 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQXuL1Dl40dZ8OHEvf:-LjR15Wd0o6P6DqcFz4d:b-fcd700) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_tier_1_2_vars.py#L229)):*
<details><summary><i>Previously, NMNS93 (Nana Mensah) wrote…</i></summary><blockquote>

I'm guessing that we capture the genome assembly from the json rather than variantCoordinates. If so, this coomment should note that we use that instead?

Also is GEL-42665 a document or a ticket? I would suggest keeping this info in a readme/somewhere else unless it's a persistent document that can be referred to. My bad if it is!
</blockquote></details>

GEL-42665 is a GeL service desk ticket. I've removed reference to it and hopefully clarified comment but let me know if needs to be clearer

**Review by andyb3 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3)* status: 0 of 5 files reviewed, 13 unresolved discussions (waiting on @aledj2, @andyb3, @NMNS93, and @woook)

___
*[100k_variant_import.py, line 8 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQYXseBTsfGc1ctE_f:-LjR5Xnn1hfQQqtT9tfp:b-hl085l) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_import.py#L8)):*
<details><summary><i>Previously, NMNS93 (Nana Mensah) wrote…</i></summary><blockquote>

> ```
> Requirements:
>     Python 2.7
>     paramiko
>     pyodbc
> ```

Could we remove this? Import statement tells me more and the shebang at the top could indicate python2.7

Standard is a requirements.txt file created with pip which we could consider making in the future.
</blockquote></details>

Have removed the requirements, agree it is clear from import statements and README what the requirements are. Also added shebangs to other two scripts. Will consider requirements.txt in future.
___
*[100k_variant_import.py, line 163 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQcgv6CBTwNr-a0U8X:-LjWG84n5QaLhqa03-CR:b-896fix) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_import.py#L163)):*
<details><summary><i>Previously, NMNS93 (Nana Mensah) wrote…</i></summary><blockquote>

There's lots happening in this function that isn't inserting the variant into Moka. It looks like the dictionary being passed to this function requires a few 'cleaning' operations before it can go in.

Can we break those out into a separate function or even class, so that the variant passed to this is ready for insert?
</blockquote></details>

Done.
___
*[100k_variant_import.py, line 172 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQd4k03knLEPkCcR2O:-LjWI48p2cLd-4X3z80C:beqidnc) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_import.py#L172)):*
<details><summary><i>Previously, NMNS93 (Nana Mensah) wrote…</i></summary><blockquote>

What's the difference between chr37/chr38 and pos37/pos38? May be worth a comment as I'm not sure what's happening
</blockquote></details>

Have added a comment but let me know if still not clear!
___
*[100k_variant_import.py, line 344 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQde8O9VHI3tSpdpPq:-LjWIn761LYCo1BI6OqO:b-896fix) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_import.py#L344)):*
<details><summary><i>Previously, NMNS93 (Nana Mensah) wrote…</i></summary><blockquote>

Small thing but please could we move this comment down to above the for loop
</blockquote></details>

Done.
___
*[100k_variant_import.py, line 432 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQe3LF2W43ke9WHVXp:-LjWLXRQDdk4axH4lCvR:b-v923on) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_import.py#L432)):*
<details><summary><i>Previously, NMNS93 (Nana Mensah) wrote…</i></summary><blockquote>

 I have seen catching all exceptions like this as a red flag before. Can it be avoided? I.e. Are the specific exceptions to catch?
</blockquote></details>

Yeah it's usually not best practice to do this, but this is basically catching any error that propagates from calling the remote script via paramiko (so any error during SSHing to server, running remote script, API calls etc. get caught by this). However whatever the error, it needs to be handled in the same way (i.e. something's gone wrong so don't import this variant) so don't think catching all will cause any problems
___
*[100k_variant_import.py, line 452 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQeUFNDJnzRgQF7zf9:-LjWKfkB-_Z_BjV-lmB2:b-vz6nbl) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_import.py#L452)):*
<details><summary><i>Previously, NMNS93 (Nana Mensah) wrote…</i></summary><blockquote>

As with previous comment. And this is also done in the else statement above?
</blockquote></details>

This has been removed

**Review by andyb3 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3)* status: 0 of 5 files reviewed, 13 unresolved discussions (waiting on @aledj2, @andyb3, @NMNS93, and @woook)

___
*[100k_variant_validator.py, line 49 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQgnyf-JfwiKZLZL_1:-LjWYc9L8VXvqSbylEUz:b-896fix) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_validator.py#L49)):*
<details><summary><i>Previously, NMNS93 (Nana Mensah) wrote…</i></summary><blockquote>

Should each variant class be initialised with the arguments that are passed to get_info?

Is get info to be used repeatedly on multiple variants? As we have all these self variables I didn't think so? Maybe the docstring can explain a bit more
</blockquote></details>

Done.
___
*[100k_variant_validator.py, line 61 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQfNpPBPUo3oluqhd5:-LjWYNI4ExktTJGGI6yP:b-cby69h) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_validator.py#L61)):*
<details><summary><i>Previously, NMNS93 (Nana Mensah) wrote…</i></summary><blockquote>

This was quite hard to follow. Potentially it could be broken up into:
* A function that gets the json
* A function that processes jsons with intergenic flags
* A function that processes jsons with gene_variant flags

get_info() could use the json flag to call the function you want. There are other ways but a dict can work too:
```python
flag_parser = {'intergenic': self.parse_intergenic, 'gene_variant': self.parse_gene_variant}
parser = flag_parser[json_flag]
parser(r.json)
```
</blockquote></details>

Done. I like the dictionary idea above!
___
*[README.md, line 111 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQhPJO8CsP1tKZURvx:-LjWYf9A3mQv1zKeio1K:b-frx3zj) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/README.md#L111)):*
<details><summary><i>Previously, NMNS93 (Nana Mensah) wrote…</i></summary><blockquote>

Requirements.txt file might be better as suggested but if not using pip I understand listing here at least!
</blockquote></details>

Yeah going to leave like this for now, but will definitely use pip for future projects!
___
*[README.md, line 117 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQh_KxEAOE0CEYqJzM:-LjWZRee45vwolps9_2E:b-896fix) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/README.md#L117)):*
<details><summary><i>Previously, NMNS93 (Nana Mensah) wrote…</i></summary><blockquote>

'are in'
</blockquote></details>

Done.

**Review by andyb3 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3)* status: 0 of 5 files reviewed, 13 unresolved discussions (waiting on @aledj2, @NMNS93, and @woook)

___
*[100k_variant_import.py, line 167 at r1](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3#-LjQbu4f4uXfE1NUWM9R:-LjW_6U3ASj0CjBWkKUV:btof5vj) ([raw file](https://github.com/moka-guys/100k_moka_variants/blob/e1326d93f8fbf31b6fcf9013abd6defa58d884d8/100k_variant_import.py#L167)):*
<details><summary><i>Previously, NMNS93 (Nana Mensah) wrote…</i></summary><blockquote>

> ```
>         Args:
>             variant = dictionary in the format output from get_additional_info() function
>         Returns:
>             status string, indicating outcome of function (e.g. imported, skipped)
> ```

Suggestion for consistency/clarity. I use the google python style guide where I can:
```
Args:
    variant(dict): Variant data output from get_additional_info().
Returns:
    status(str): Indicates outcome of variant import into moka. Examples: imported, skipped.
```

</blockquote></details>

Like this idea. Will definitely use for future projects and think we should adopt this as a team. I'll probably leave the docstrings in here as is for now if that's OK (just because there are a lot to update!) but will stick an issue in to address this

**Review by NMNS93 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/moka-guys/100k_moka_variants/3)* status: 0 of 5 files reviewed, all discussions resolved (waiting on @aledj2 and @woook)
